## try - catch문
### try블록에는 예외가 발생할 가능성이 있는 코드를 작성하고, try블록 안에서 예외가 발생하는 경우 catch블록이 수행됨
```java
try {
예외가 발생할 수 있는 코드 부분
}catch(처리할 예외타입 e) {
try 블록 안에서 예외가 발생했을때 예외를 처리하는 부분
}
```

## ## try - catch문 예제
```java
package ch09;
//배열의 인덱스 크기 초과 했을때, 오류 잡기
public class ArrayIndexException {
	
	public static void main(String[] args) {
		
		int[] arr = {1,2,3,4,5};
		
		try {
			
			for(int i = 0; i <= 5; i ++) {
				System.out.println(arr[i]);
			}
		}catch(ArrayIndexOutOfBoundsException e) {
			System.out.println(e.getMessage());
			System.out.println(e.toString());
		}
		//try ~ catch 사용을 해서 비록 오류는 발생했지만, 비정상종료가 안되서 here!!!이 찍힌다.
		System.out.println("here!!!!");
		
	}

}
```

## try-catch-finnaly 문
### finally 블럭에서 파일을 닫거나 네트웍을 닫는 리소스 해제 구현을 함
### try{} 블럭이 수행되는 경우, finally{} 블럭은 항상 수행 됨
### 여러 개의 예외 블럭이 있는 경우 각각에서 리소스를 해제하지 않고, finally 블록에서 해제하도록 구현함<br></br>

## 컴파일러에 의해 예외가 처리 되는 예(파일 에러 처리)
### try에서 파일이 없는 경우, 파일이 없고 하니까 해제를 해줘야하 하는대, 그부분을 finally에서 동작하게 하는 것이다

```java
package ch09;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;

public class FileExceptionHandling {

	public static void main(String[] args) {

		//FileInputStream 클래스
		FileInputStream fis = null;
		try {
			fis  = new FileInputStream("a.txt");
			System.out.println("read");
						
		} catch (FileNotFoundException e) {
			System.out.println(e);
		} finally {
			if(fis != null) { 
				try {
					fis.close(); //파일을 닫아야 함
				} catch (IOException e) {
					
					e.printStackTrace();
				}
				
			}
			System.out.println("finally는 항상 불려진다 ");
			
		}
		System.out.println("end");
	}

}
```

## try-with-resources 문
### 리소스를 사용하는 경우 close() 하지 않아도 자동으로 해제 되도록 함
### 자바 7부터 제공되는 구문
### 리소스를 try() 내부에서 선언해야 함
### 해당 리소스 클래스가 AutoCloseable 인터페이스를 구현 해야 함
### FileInputStream의 경우에는 AutoCloseable을 구현하고 있음

```java
package ch09;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;

public class FileExceptionHandling {

	public static void main(String[] args) {

		//FileInputStream 클래스
		
		try(FileInputStream fis = new FileInputStream("a.txt")) {
			//FileInputStream fis = new FileInputStream("a.txt") 이부분에서 자동으로 close 된다
			System.out.println("read");
						
		} catch (FileNotFoundException e) {
			//파일이 없을때 나는 에러ㅏ -> FileNotFoundException
			e.printStackTrace();
		} catch (IOException e) {
			//IOException close할때 나는 에러 
			e.printStackTrace();
		} 
		System.out.println("end");
	}

}
```

## AutoCloseable 예제
### 1. AutoCloseable을 구현 한 AutoCloseableObj 인터페이스 생성
```java
package ch09;

public class AutoCloseableObj implements AutoCloseable {

	@Override
	public void close() throws Exception {ArrayIndexException

		System.out.println("closeing......");
	}

}
```
### 2. AutoCloseTest 클래스 생성
```java
package ch09;

public class AutoCloseTest {

	public static void main(String[] args) {

		AutoCloseableObj obj = new AutoCloseableObj();
		
		try(obj) {
			//exception 상황일 경우
			//강제로 부를거다 -> throw을 쓰면 Exception 강제로 발생함
			throw new Exception();
		}catch(Exception e) {
			System.out.println("exception");
		}
		System.out.println("end");
		
		
			
			
		
	}

}
```
# 예외처리 미루기
### 예외처리는 예외가 발생하는 문자에서 try-catch 블록으로 처리하는 방법, 이를 사용하는 부분에서 처리하는 방법이 있다
### throws를 이용하면 예외가 발생할 수 있는 부분을 사용하는 문장에서 예외를 처리할 수 있음


